src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2626,210 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+{
+    auto hasCheckedOverriddenEquals = [](MethodTable* mt) -> BOOL {
+        return mt->HasCheckedOverriddenEquals();
+    };
+    auto setHasCheckedOverriddenEquals = [](MethodTable* mt) -> void {
+        mt->SetHasCheckedOverriddenEquals();
+    };
+    auto hasOverriddenEquals = [](MethodTable* mt) -> BOOL {
+        return mt->HasOverriddenEquals();
+    };
+    auto setHasOverriddenEquals = [](MethodTable* mt) -> void {
+        mt->SetHasOverriddenEquals();
+    };
+
+    if (g_pValueTypeMT == NULL)
+    {
+        g_pValueTypeMT = MscorlibBinder::GetClass(CLASS__VALUE_TYPE);
+    }
+
+    if (g_slotEquals == 0)
+    {
+        g_slotEquals = MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__EQUALS)->GetSlot();
+    }
+
+    return HasOverriddenMethod(mt,
+                               g_pValueTypeMT,
+                               g_slotEquals,
+                               hasCheckedOverriddenEquals,
+                               setHasCheckedOverriddenEquals,
+                               hasOverriddenEquals,
+                               setHasOverriddenEquals);
+}
+
+static BOOL HasOverriddenGetHashCode(MethodTable* mt)
+{
+    auto hasCheckedOverriddenGetHashCode = [](MethodTable* mt) -> BOOL {
+        return mt->HasCheckedOverriddenGetHashCode();
+    };
+    auto setHasCheckedOverriddenGetHashCode = [](MethodTable* mt) -> void {
+        mt->SetHasCheckedOverriddenGetHashCode();
+    };
+    auto hasOverriddenGetHashCode = [](MethodTable* mt) -> BOOL {
+        return mt->HasOverriddenGetHashCode();
+    };
+    auto setHasOverriddenGetHashCode = [](MethodTable* mt) -> void {
+        mt->SetHasOverriddenGetHashCode();
+    };
+
+    if (g_pValueTypeMT == NULL)
+    {
+        g_pValueTypeMT = MscorlibBinder::GetClass(CLASS__VALUE_TYPE);
+    }
+
+    if (g_slotGetHashCode == 0)
+    {
+        g_slotGetHashCode = MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__GET_HASH_CODE)->GetSlot();
+    }
+
+    return HasOverriddenMethod(mt,
+                               g_pValueTypeMT,
+                               g_slotGetHashCode,
+                               hasCheckedOverriddenGetHashCode,
+                               setHasCheckedOverriddenGetHashCode,
+                               hasOverriddenGetHashCode,
+                               setHasOverriddenGetHashCode);
+}
+
+// Return true if the valuetype does not contain pointer, is tightly packed, 
+// does not have floating point number field and does not override Equals method.
 FCIMPL1(FC_BOOL_RET, ValueTypeHelper::CanCompareBits, Object* obj)
 {
     FCALL_CONTRACT;
 
     _ASSERTE(obj != NULL);
     MethodTable* mt = obj->GetMethodTable();
-    FC_RETURN_BOOL(!mt->ContainsPointers() && !mt->IsNotTightlyPacked());
+    if (mt->ContainsPointers()
+        || mt->IsNotTightlyPacked()
+        || mt->HasFloatingPointNumberField())
+    {
+        FC_RETURN_BOOL(FALSE);
+    }
+
+    FC_RETURN_BOOL(!HasOverriddenEquals(mt));
```

Comment:
The code that iterates the fields, etc. has to be executed under `HELPER_METHOD_FRAME`. 

https://github.com/dotnet/coreclr/blob/master/Documentation/botr/mscorlib.md and https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md describes the rules for writing FCalls and manually managed code.